src/Common/src/System/IO/ReadOnlyMemoryStream.cs

```diff
@@ -0,0 +1,154 @@
+        {
+            long pos =
+                origin == SeekOrigin.Begin ? offset :
+                origin == SeekOrigin.Current ? _position + offset :
+                origin == SeekOrigin.End ? _content.Length + offset :
+                throw new ArgumentOutOfRangeException(nameof(origin));
+
+            if (pos > int.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            else if (pos < 0)
+            {
+                throw new IOException(SR.IO_SeekBeforeBegin);
+            }
+
+            _position = (int)pos;
+            return _position;
+        }
+
+        public override int ReadByte()
+        {
+            ReadOnlySpan<byte> s = _content.Span;
+            return _position < s.Length ? s[_position++] : -1;
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateReadArrayArguments(buffer, offset, count);
+            return Read(new Span<byte>(buffer, offset, count));
+        }
+
+        public override int Read(Span<byte> destination)
+        {
+            int remaining = _content.Length - _position;
+
+            if (remaining <= 0 || destination.Length == 0)
+            {
+                return 0;
+            }
+            else if (remaining <= destination.Length)
+            {
+                _content.Span.Slice(_position).CopyTo(destination);
+                _position = _content.Length;
+                return remaining;
+            }
+            else
+            {
+                _content.Span.Slice(_position, destination.Length).CopyTo(destination);
+                _position += destination.Length;
+                return destination.Length;
+            }
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ValidateReadArrayArguments(buffer, offset, count);
+            return cancellationToken.CanBeCanceled ?
+                Task.FromCanceled<int>(cancellationToken) :
+                Task.FromResult(Read(new Span<byte>(buffer, offset, count)));
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default(CancellationToken)) =>
+            cancellationToken.CanBeCanceled ?
+                new ValueTask<int>(Task.FromCanceled<int>(cancellationToken)) :
+                new ValueTask<int>(Read(destination.Span));
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state) =>
+            TaskToApm.Begin(ReadAsync(buffer, offset, count), callback, state);
+
+        public override int EndRead(IAsyncResult asyncResult) =>
+            TaskToApm.End<int>(asyncResult);
+
+        public override void CopyTo(Stream destination, int bufferSize)
+        {
+            if (_content.Length > _position)
```

Comment:
Should we do arg validation here?  (We do it everywhere else)

Reply:
Yup.